builder/googlecompute/block_device.go

```diff
@@ -305,19 +317,32 @@ func (bd BlockDevice) generateDiskAttachment() *compute.AttachedDisk {
 			Mode:              bd.AttachmentMode,
 			Type:              "SCRATCH",
 			InitializeParams: &compute.AttachedDiskInitializeParams{
-				DiskType: fmt.Sprintf("zones/%s/diskTypes/local-ssd", bd.zone),
+				DiskType:    fmt.Sprintf("zones/%s/diskTypes/local-ssd", bd.zone),
+				SourceImage: bd.SourceImage,
 			},
 		}
 	}
 
-	return &compute.AttachedDisk{
+	disk := &compute.AttachedDisk{
 		AutoDelete:        bd.shouldAutoDelete(),
 		Boot:              false,
 		DeviceName:        bd.DeviceName,
 		Interface:         bd.InterfaceType,
 		Mode:              bd.AttachmentMode,
 		DiskEncryptionKey: bd.DiskEncryptionKey.ComputeType(),
 		Type:              "PERSISTENT",
-		Source:            bd.SourceVolume,
 	}
+
+	if bd.SourceImage != "" {
+		// `DiskName` here needs to match `compute.Disk` created above
+		// otherwise Google creates another disk.
+		disk.InitializeParams = &compute.AttachedDiskInitializeParams{
+			DiskName:    bd.DiskName,
+			SourceImage: bd.SourceImage,
+		}
+	} else {
+		disk.Source = bd.SourceVolume
```

Comment:
Since the `SourceVolume` and the `SourceImage` are mutually exclusive, I believe this can remain part of the definition for the `AttachedDisk`?